### PR TITLE
Unit tests for MultiTenantTransformationFilterTest

### PR DESCRIPTION
**Pull request by k-wall**
The POC didn't contribute any unit tests for the multitenant filter.
I'd like to fix that.

### CODE REVIEW

**Comment by k-wall** [comment]
Constructing the the kafka value objects in a fluent concise manner doesn't seem to be terribly easy.   The best way I've found so far is to lean on the JSON request/response transformation abilities of the Kafka itself.   Doing that gives a parameterized test, which I think, remains readable.

What do you think to the approach?

**Comment by robobario** [comment]
Maybe the JSON blobs could be externalised? Like have some path in `src/test/resources` that contains pairs of files like `request-create-topic-in.json` and `request-create-topic-expected-diff.json`. This might make it easier to manipulate them in bulk later since you could operate on them with `jq` or language-of-choice if you needed to update lots of inputs or assertions in future.

Also for unit level I'd rather have the request and response methods checked separately, despite them forming a logical pair in reality. So I know that one test is checking one request or one response.

**Comment by robobario** [comment]
Also I can see there is an argument for parameterisation as we are checking for topicname prefixing in each case, but practically I find that argument array hard to read and wonder if we are better off splitting it into individual methods.

**Comment by k-wall** [comment]
> Maybe the JSON blobs could be externalised? 

+1 to the json file idea, though I think having the request/response along side the expected diff might be best.  I'll try that refactoring.  

> Also for unit level I'd rather have the request and response methods checked separately, despite them forming a logical pair in reality. So I know that one test is checking one request or one response.

I think that request and response test do belong together.  Request and response formats have evolved together (e.g. the evolution towards topic ids), so I think having a single file expressing both halves will make for easier developer comprehension.

I am thinking of a JSON file with an object structured like this:

```yaml
[
   - testName: 
     version
     request:
       payload:
       diff:
     response:
       payload:
       diff:
]
```

and have a file per RPC.

> Also I can see there is an argument for parameterisation as we are checking for topicname prefixing in each case, but practically I find that argument array hard to read and wonder if we are better off splitting it into individual methods.
> edit: I imagine this can be improved, but currently if a test failed it won't be immediately clear which case has failed.

I think the tooling should make it clear which parameter is failure and the IDE let you re-run invoke a single instance of a parameterized test.

**Comment by robobario** [comment]
> I think the tooling should make it clear which parameter is failure and the IDE let you re-run invoke a single instance of a parameterized test.

Yes, but the parameters could be unreadable. In the usage so far it works since you see `apiMessageType=DELETE` in the failure message you'll know roughly file a failure came from. But if we make this technique available to other filter authors it may be they have lots of similar looking cases for the same message type with different input. If we are going for an array of scenarios per file I think it would help understanding if we can add a description to each scenario that shows up somewhere in the error.

**Comment by k-wall** [comment]
> > I think the tooling should make it clear which parameter is failure and the IDE let you re-run invoke a single instance of a parameterized test.
> 
> Yes, but the parameters could be unreadable. In the usage so far it works since you see something like: `apiMessageType=DELETE` in the failure message you'll know roughly which file a failure came from.

I did have thoughts about either having the yaml carry a user-defined test name, or exposing the test case filename/array index as parameterized test parameters.

> But if we make this technique available to other filter authors it may be they have lots of similar looking cases for the same message type with different input. If we are going for an array of scenarios per file I think it would help understanding if we can add an optional description to each scenario that shows up somewhere in the error.

I vote we wait until we discover other filter use-cases with similar test requirements before making the test mechanism part of a testing API.  I'm sure we'll want to rev the test mechanism as we build out multitenant functions.  Let's not abstract too early.

**Review by robobario — APPROVED** [review]
LGTM

**Comment by robobario** [comment]
Note: @SamBarker found that tests weren't running in IDEA when using `Import using the new IntelliJ Workspace Model API (experimental)`. You see:
```
org.junit.platform.commons.PreconditionViolationException: Configuration error: You must configure at least one set of arguments for this @ParameterizedTest
```
when the tests are run.

**Comment by k-wall** [comment]
> Note: @SamBarker found that tests weren't running in IDEA when using `Import using the new IntelliJ Workspace Model API (experimental)`. You see:
> 
> ```
> org.junit.platform.commons.PreconditionViolationException: Configuration error: You must configure at least one set of arguments for this @ParameterizedTest
> ```
> 
> when the tests are run.

Yeah I've noticed that too.  It seems that Intellij isn't proper setting up the classpath to include resources when in that mode.

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    private record RequestResponseTestDef(String source, ApiMessageType apiKey, RequestHeaderData header, ApiMessageTestDef request, ApiMessageTestDef response) {
+    }
+
+    private record Converters(BiFunction<JsonNode, Short, ApiMessage> requestReader,
+                              BiFunction<JsonNode, Short, ApiMessage> responseReader,
+                              BiFunction<ApiMessage, Short, JsonNode> requestWriter,
+                              BiFunction<ApiMessage, Short, JsonNode> responseWriters) {
+    }
+
+    private static final Map<ApiMessageType, Converters> converters = Map.of(
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java:98]: If convenient access to the DataJsonConverters is something that is useful elsewhere, then we can factor it out. Let's wait and see what testing patterns emerge.

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+            var request = readerFunc.apply(node.get("payload"), apiVersion);
+            var requestDiff = node.get("diff");
+            return new ApiMessageTestDef(request, requestDiff);
+        }
+        return null;
+    }
+
+    private final MultiTenantTransformationFilter filter = new MultiTenantTransformationFilter();
+
+    private final KrpcFilterContext context = mock(KrpcFilterContext.class);
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java:197]: We are already using Mockito without the MockitoJunit5Extension elsewhere, so I have just followed established usage patterns.  Personally I don't feel strongly about which way it is used, providing we do it consistently.

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+    private final ArgumentCaptor<ApiMessage> apiMessageCaptor = ArgumentCaptor.forClass(ApiMessage.class);
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java:199]: We are already using Mockito without the MockitoJunit5Extension elsewhere, so I have just followed established usage patterns.

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        when(context.sniHostname()).thenReturn("tenant1.kafka.example.com");
+    }
+
+    public static Stream<Arguments> requests() throws Exception {
+        return requestResponseTestDefinitions().map(t -> Arguments.of(t.source(), t.apiKey(), t.header(), t.request()));
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    void requests(@SuppressWarnings("unused") String source, ApiMessageType apiMessageType, RequestHeaderData header, ApiMessageTestDef requestTestDef) {
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java:212]: Renamed

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                    (o, ver) -> OffsetCommitRequestDataJsonConverter.write(((OffsetCommitRequestData) o), ver),
+                    (o, ver) -> OffsetCommitResponseDataJsonConverter.write(((OffsetCommitResponseData) o), ver)),
+            ApiMessageType.FETCH, new Converters(
+                    FetchRequestDataJsonConverter::read,
+                    FetchResponseDataJsonConverter::read,
+                    (o, ver) -> FetchRequestDataJsonConverter.write(((FetchRequestData) o), ver),
+                    (o, ver) -> FetchResponseDataJsonConverter.write(((FetchResponseData) o), ver)));
+
+    private static Stream<RequestResponseTestDef> requestResponseTestDefinitions() throws Exception {
+        return ClassPath.from(MultiTenantTransformationFilterTest.class.getClassLoader()).getResources().stream()
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java:146]: added a guard and advice to check the IDEA Test Runner's classopath.

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java:212]: Yes, the intent it to allow the identification of the test.  I've made it more explicit.


Here's how it looks in Intellij:

Screenshot 2023-03-15 at 13 19 56


The XML artefact generated by Surefire includes the display name too:

Screenshot 2023-03-15 at 13 23 51

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        }
+    }
+
+    private static RequestResponseTestDef buildRequestResponseTestDef(JsonNode jsonNode) {
+        var apiMessageType = jsonNode.get("apiMessageType");
+        var version = jsonNode.get("version");
+        var source = jsonNode.get("source");
+        assertNotNull(apiMessageType);
+        assertNotNull(version);
+        assertNotNull(source);
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java]: Done

**Comment by k-wall** [comment]
> > Note: @SamBarker found that tests weren't running in IDEA when using `Import using the new IntelliJ Workspace Model API (experimental)`. You see:
> > ```
> > org.junit.platform.commons.PreconditionViolationException: Configuration error: You must configure at least one set of arguments for this @ParameterizedTest
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > when the tests are run.
> 
> Yeah I've noticed that too. It seems that Intellij isn't properly setting up the classpath to include resources when in that mode.

I spent a bit more time looking into this without making much progress.   The issue is Intellij doesn't copy the test-resources to the test-classes, so the resources don't end up on the classpath.  You can fix by doing this.

```bash
 mvn process-test-resources -f kroxylicious-multitenant
```

* I tried setting up a new multi module maven project in the  IntelliJ Workspace Model  mode and couldn't reproduce the issue. 
* I also tried cleaning kroxylicious's Intellij files away and reimporting from scratch. No help.
* I reproduced the problem in other modules within kroxylicious, so I confident that issue isn't uniqie to the multi-tenant model.

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java]: Ok cool. Tempted to include as a customised display name but that's polish for another day.

[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java]: Ignore me. Thought I had updated my copy of the PR in IntelliJ but that hadn't pulled down the latest changes.

File: kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java [hunk 7]
```diff
@@ -204,12 +216,12 @@ public void beforeEach() {
...
     public static Stream<Arguments> requests() throws Exception {
-        return requestResponseTestDefinitions().map(t -> Arguments.of(t.source(), t.apiKey(), t.header(), t.request()));
+        return requestResponseTestDefinitions().map(t -> Arguments.of(t.testName(), t.apiKey(), t.header(), t.request()));
     }
 
-    @ParameterizedTest
+    @ParameterizedTest(name = "{0}")
     @MethodSource
-    void requests(@SuppressWarnings("unused") String source, ApiMessageType apiMessageType, RequestHeaderData header, ApiMessageTestDef requestTestDef) {
+    void requests(@SuppressWarnings("unused") String testName, ApiMessageType apiMessageType, RequestHeaderData header, ApiMessageTestDef requestTestDef) {
```
Comment (SamBarker) [inline kroxylicious-multitenant/src/test/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilterTest.java:225]: nit: `requests` -> `requestsTransformed`

Though I'd be tempted to leave `Stream<Arguments> requests()` and use `@MethodSource("requests")` here

**Review by SamBarker — APPROVED** [review]
LGTM

Thanks @k-wall

**Pull request merged**